src/utils/interpolate.hpp

```diff
@@ -0,0 +1,134 @@
+#ifndef INTERPOLATE_HPP_
+#define INTERPOLATE_HPP_
+//========================================================================================
+// Parthenon performance portable AMR framework
+// Copyright(C) 2022 The Parthenon collaboration
+// Licensed under the 3-clause BSD License, see LICENSE file for details
+//========================================================================================
+//! \file interpolate.hpp
+//  \brief Interpolation using cubic Hermite polynomials
+//
+// Performs monotone interpolation using cubic Hermite polynomials.
+// Reference: "Monotone piecewise cubic interpolation," SIAM J. Numer. Anal.,
+//            Vol. 17, No. 2, April 1980. [https://doi.org/10.1137/0717021]
+//========================================================================================
+
+#include "../kokkos_abstraction.hpp"
+#include <ostream>
+
+template <class VectorContainer>
+class MonotoneInterpolator {
+ public:
+  using Real = typename VectorContainer::value_type;
+  MonotoneInterpolator(VectorContainer const &x, VectorContainer const &y);
+  KOKKOS_FUNCTION KOKKOS_FORCEINLINE_FUNCTION
+  MonotoneInterpolator(MonotoneInterpolator const &rhs)
+      : x_min_(rhs.x_min_), x_max_(rhs.x_max_), x_vec_(rhs.x_vec_), f_vec_(rhs.f_vec_),
+        d_vec_(rhs.d_vec_) {}
+
+  KOKKOS_FUNCTION KOKKOS_FORCEINLINE_FUNCTION auto operator()(Real x) const -> Real;
+  KOKKOS_FUNCTION KOKKOS_FORCEINLINE_FUNCTION auto min() const -> Real { return x_min_; }
+  KOKKOS_FUNCTION KOKKOS_FORCEINLINE_FUNCTION auto max() const -> Real { return x_max_; }
+  auto ConstructVectorContainer(size_t size) -> VectorContainer;
+
+ private:
+  Real x_min_{};
+  Real x_max_{}; // x_min and x_max are the min and max of x_vec_
+  VectorContainer x_vec_{};
```

Comment:
Are these assumed to have reference semantics?

Reply:
I don't think so... Can you provide an example of how this could cause trouble?

Reply:
you're constructing `x_vec_` by copy when you call the interpolator, so you're copying around the data unless `VectorContainer` has refernece semantics. That may be a feature or a bug. If `VectoContainer` is always a `Kokkos::View`, then we're fine though.